lib/containers/accumulators/review-threads-accumulator.js

```diff
@@ -0,0 +1,164 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {graphql, createPaginationContainer} from 'react-relay';
+
+import {PAGE_SIZE, PAGINATION_WAIT_TIME_MS} from '../../helpers';
+import {RelayConnectionPropType} from '../../prop-types';
+import Accumulator from './accumulator';
+import ReviewCommentsAccumulator from './review-comments-accumulator';
+
+export class BareReviewThreadsAccumulator extends React.Component {
+  static propTypes = {
+    // Relay props
+    relay: PropTypes.shape({
+      hasMore: PropTypes.func.isRequired,
+      loadMore: PropTypes.func.isRequired,
+      isLoading: PropTypes.func.isRequired,
+    }).isRequired,
+    pullRequest: PropTypes.shape({
+      reviewThreads: RelayConnectionPropType(
+        PropTypes.object,
+      ),
+    }),
+
+    // Render prop. Called with (array of errors, array of threads, map of comments per thread, loading)
+    children: PropTypes.func.isRequired,
+
+    // Called right after refetch happens
+    onDidRefetch: PropTypes.func.isRequired,
+  }
+
+  render() {
+    const resultBatch = this.props.pullRequest.reviewThreads.edges.map(edge => edge.node);
+    return (
+      <Accumulator
+        relay={this.props.relay}
+        resultBatch={resultBatch}
+        onDidRefetch={this.props.onDidRefetch}
+        pageSize={PAGE_SIZE}
+        waitTimeMs={PAGINATION_WAIT_TIME_MS}>
+        {this.renderReviewThreads}
+      </Accumulator>
+    );
+  }
+
+  renderReviewThreads = (err, threads, loading) => {
+    if (err) {
+      return this.props.children({
+        errors: [err],
+        commentThreads: [],
+        loading,
+      });
+    }
+
+    return this.renderReviewThread({errors: [], commentsByThread: new Map(), loading}, threads);
+  }
+
+  renderReviewThread = (payload, threads) => {
+    if (threads.length === 0) {
+      const commentThreads = [];
+      payload.commentsByThread.forEach((comments, thread) => {
+        commentThreads.push({thread, comments});
+      });
+      return this.props.children({
+        commentThreads,
+        errors: payload.errors,
+        loading: payload.loading,
+      });
+    }
+
+    const [thread] = threads;
+    return (
+      <ReviewCommentsAccumulator
+        onDidRefetch={this.props.onDidRefetch}
+        reviewThread={thread}>
+        {({error, comments, loading: threadLoading}) => {
+          if (error) {
+            payload.errors.push(error);
+          }
+          payload.commentsByThread.set(thread, comments);
+          payload.loading = payload.loading || threadLoading;
+          return this.renderReviewThread(payload, threads.slice(1));
```

Comment:
Recursion FTW!

![recursive gif](https://media.giphy.com/media/loerhc8yp0imk/giphy.gif)

Reply:
Okay, that .gif is amazing. That's getting saved for later